BUG: non-integers can end up in dtype offsets

ahaldane opened the pull request:
Fix is to convert offsets to python ints at dtype creation.
Fixes #8059

ahaldane committed:
BUG: non-integers can end up in dtype offsets

Fix is to convert offsets to python ints at dtype creation.
Fixes #8059

seberg commented on numpy/core/src/multiarray/descriptor.c:
```diff
@@ -1072,7 +1072,13 @@ _convert_from_dict(PyObject *obj, int align)
             }
             offset = PyInt_AsLong(off);
-            PyTuple_SET_ITEM(tup, 1, off);
+            if (PyErr_Occurred()) {
```
`if ((offset == -1) && PyErr_Occured()) {` avoids the check most of the time I think.

seberg replied:
Two rather unrelated side points I am not sure about:
 1. What about floats, etc.? Maybe it would make sense to use the Index conversion stuff?
 2. Where do we check `int` size? We only support integer sized offsets (since dtype is limited to int).

ahaldane replied:
what is the index conversion stuff? I don't think I'm familiar and searching didn't help.

ahaldane replied:
also `PyInt_AsLong` takes care of floats etc

seberg replied:
I mean our own (or the python version we mostly call there) NpyPyIntAsIntp (or whate it was), which specificially does not allow floats.

ahaldane commented:
Got it, and updated. 

I also added a check that `offset` is positive - previously negative was allowed and I could read far back in memory! Also I made the conversion of `itemsize` safer.

seberg commented:
Nice. Hmmmpff, I may have thought too little about my comments, though. At least if we want to backport this, the more strict integer conversion could be a regression (though maybe it cannot be, because this is almost always only called indirectly).

ahaldane commented:
Anyone trying to use weird non-integer offsets would probably cause numpy to blow up in other ways ...probably.

Also, I checked and #8059 is not a new bug I introduced, it is old. I checked in 1.8.0 and 1.10.0 and the examples raise exceptions back then too.

Personally I would not worry about a backport, I think, since this is such an old and obscure oversight.

seberg commented:
OK, then I think I am happy. Thanks.
